service/controller/resource/nodecontroller/create.go

```diff
@@ -0,0 +1,84 @@
+package nodecontroller
+
+import (
+	"context"
+
+	"github.com/giantswarm/apiextensions/v3/pkg/apis/provider/v1alpha1"
+	workload "github.com/giantswarm/errors/tenant"
+	"github.com/giantswarm/k8sclient/v5/pkg/k8sclient"
+	"github.com/giantswarm/microerror"
+	workloadcluster "github.com/giantswarm/tenantcluster/v4/pkg/tenantcluster"
+
+	"github.com/giantswarm/kvm-operator/service/controller/internal/nodecontroller"
+	"github.com/giantswarm/kvm-operator/service/controller/key"
+)
+
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	cr, err := key.ToCustomObject(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	r.logger.Debugf(ctx, "ensuring node controller is running for workload cluster")
+
+	if _, ok := r.controllers[nodeControllerKey(cr)]; !ok {
```

Comment:
I'm not sure if I should recreate the controller on every reconciliation. I'm worried that certs, for example, could change and the controller would be silently failing.

Reply:
I am a bit worried about dynamically creating these controllers. From the current logic implemented it looks reasonably safe. I could see it causing accidental memory leaks easily though.

Booting the controllers on each reconciliation could create an unreasonable amount of logs.

Reply:
I have added some very careful code to ensure that we would only reboot a controller if the certs or KVMConfig spec changes and to make sure that only one controller per cluster can run at one time. Let me know what you think of the logic if you have time to check it out.

Reply:
Also I'll take a look at the logs to make sure it's still readable.